ChemistryLib/PhreeqcKernel.cpp

```diff
@@ -0,0 +1,220 @@
+    setAqueousSolutions(process_solutions);
+    setTimeStep(dt);
+
+    execute(process_solutions);
+}
+
+void PhreeqcKernel::setAqueousSolutions(
+    std::vector<GlobalVector*> const& process_solutions)
+{
+    // Loop over chemical systems
+    auto const num_chemical_systems = process_solutions[0]->size();
+    for (auto chemical_system_id = 0; chemical_system_id < num_chemical_systems;
+         ++chemical_system_id)
+    {
+        auto& aqueous_solution = Rxn_solution_map[chemical_system_id];
+
+        // Get or create initial aqueous solution
+        if (!aqueous_solution.Get_initial_data())
+        {
+            aqueous_solution.Set_initial_data(
+                &_templated_initial_aqueous_solution);
+            aqueous_solution.Set_new_def(true);
+        }
+        auto initial_aqueous_solution = aqueous_solution.Get_initial_data();
+
+        auto& components = initial_aqueous_solution->Get_comps();
+        // Loop over transport process id map to retrieve component
+        // concentrations from process solutions or to update process
+        // solutions after chemical calculation by Phreeqc
+        for (auto const& process_id_to_master_pair : _process_id_to_master_map)
+        {
+            auto& transport_process_id = process_id_to_master_pair.first;
+            auto& transport_process_solution =
+                process_solutions[transport_process_id];
+
+            auto& master_species = process_id_to_master_pair.second;
+            auto& element_name = master_species->elt->name;
+            if (strcmp(element_name, "H") == 0)
+            {
+                // Set pH value by hydrogen concentration.
+                double const pH = -std::log10(
+                    transport_process_solution->get(chemical_system_id));
+                aqueous_solution.Set_ph(pH);
+            }
+            else
+            {
+                // Set component concentrations.
+                auto const& concentration =
+                    transport_process_solution->get(chemical_system_id);
+                components[element_name].Set_input_conc(concentration);
+            }
+        }
+    }
+}
+
+void PhreeqcKernel::setTimeStep(double const dt)
+{
+    // Loop over rxn kinetics map
+    for (auto& rxn_kinetics : Rxn_kinetics_map)
+    {
+        auto& kinetics = rxn_kinetics.second;
+        kinetics.Get_steps().push_back(dt);
+    }
+}
+
+void PhreeqcKernel::execute(std::vector<GlobalVector*>& process_solutions)
+{
+    auto const num_chemical_systems = process_solutions[0]->size();
+    for (auto chemical_system_id = 0; chemical_system_id < num_chemical_systems;
+         ++chemical_system_id)
+    {
+        Rxn_new_solution.insert(chemical_system_id);
+        new_solution = 1;
+        use.Set_n_solution_user(chemical_system_id);
+
+        if (!Rxn_kinetics_map.empty())
+        {
+            use.Set_kinetics_in(true);
+            use.Set_n_kinetics_user(chemical_system_id);
+        }
+
+        pr.all = false;
```

Comment:
What is `pr`?

Reply:
Defined by third-party library Phreeqc. Moved into a private class method.